### PR TITLE
switch releasing based on tag

### PR DESCRIPTION
**Pull request by radTuti**
## Description

This creates a separate semaphore file to use for releasing Operator based off tagging

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Review by tmjd — CHANGES_REQUESTED** [review]
It seems like we're changing how we're cleaning the cache to make the building of a tag quicker with the result of slowing down any re-runs of a failed run. I would guess re-runs are going to happen much more often than tag builds. So that seems like a bad trade off.

Also does this mean if someone tagged before the Build ran that this would fail?

Seems like a tag should trigger the main operator Semaphore pipeline which should then have a promotion of the 'release.yaml'  pipeline if it is on a tag.

File: Makefile [hunk 1]
```diff
@@ -501,9 +501,15 @@ endif
 ###############################################################################
 # Release
 ###############################################################################
-## Determines if we are on a tag and if so builds a release.
-maybe-build-release:
-	./hack/maybe-build-release.sh
+VERSION_REGEX := ^v?[0-9]+\.[0-9]+\.[0-9]+(-[a-zA-Z0-9]+(\.[a-zA-Z0-9]+)*)?(\+[a-zA-Z0-9]+(\.[a-zA-Z0-9]+)*)?$$
```
[the remark this answers is not in the document]
Comment (radTuti) [reply Makefile:504]: I tried to be as expansive to accomodate the recent cloud-only operator but I will keep it to the expected if that is fine.

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -162,6 +162,10 @@ promotions:
     pipeline_file: push_images.yml
     auto_promote:
       when: "branch =~ 'master|release-.*'"
+  - name: Release
+    pipeline_file: release.yml
+    auto_promote:
+      when: "tag =~ '^v'"
```
Comment (tmjd) [inline .semaphore/semaphore.yml:168]: Looking at the semaphore docs it looks like this should be `result = 'passed' AND tag =~ '^v1\.'`  according to the example [here](https://docs.semaphoreci.com/using-semaphore/promotions#tagged)

Comment (radTuti) [reply .semaphore/semaphore.yml:168]: right!

File: Makefile [hunk 3]
```diff
@@ -501,9 +501,20 @@ endif
...
+	$(eval VALID_TAG := $(shell echo $(RELEASE_TAG) | grep -Eq "$(VERSION_REGEX)" && echo true))
+	$(if $(VALID_TAG),,$(error $(RELEASE_TAG) is not a valid version. Please use a version in the format vX.Y.Z))
+
+# Skip releasing if the image already exists.
+	@if !$(MAKE) VERSION=$(RELEASE_TAG) release-check-image-exists; then \
+		echo "Images for $(RELEASE_TAG) already exists"; \
+		exit 0; \
+	fi
+
+	$(MAKE) release release-publish-images VERSION=$(RELEASE_TAG)
```
Comment (tmjd) [inline Makefile:515]: It might make it easier to debug if these were on separate lines.

**Review by tmjd — APPROVED** [review]
One minor comment that you can ignore if you don't think it would help.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, squash the commits when I merge it, and delete the branch after I've merged it.

**Comment by marvin-tigera** [comment]
I'm sorry but I failed delete the branch after merging the pull request.

**Pull request merged**